packages/auth/src/model/password_policy.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PasswordPolicy, PasswordValidationStatus } from "./public_types";
+
+/**
+ * Custom strength options supported in the password policy.
+ */
+export interface PasswordPolicyCustomStrengthOptions {
+  /**
+   * Minimum password length.
+   */
+  readonly minPasswordLength?: number;
+  /**
+   * Maximum password length.
+   */
+  readonly maxPasswordLength?: number;
+  /**
+   * Whether the password should contain a lowercase letter.
+   */
+  readonly containsLowercaseLetter?: boolean;
+  /**
+   * Whether the password should contain an uppercase letter.
+   */
+  readonly containsUppercaseLetter?: boolean;
+  /**
+   * Whether the password should contain a numeric character.
+   */
+  readonly containsNumericCharacter?: boolean;
+  /**
+   * Whether the password should contain a non-alphanumeric character.
+   */
+  readonly containsNonAlphanumericCharacter?: boolean;
+}
+
+/**
+ * Internal typing of password policy that includes the schema version and methods for
+ * validating that a password meets the policy. The developer does not need access to
+ * these properties and methods, so they are excluded from the public typing.
+ *
+ * @internal
+ */
+export interface PasswordPolicyInternal extends PasswordPolicy {
+  /**
+   * Schema version of the password policy.
+   */
+  readonly schemaVersion: number;
+  /**
+   * Validates the password against the policy.
+   * @param password Password to validate.
+   */
+  validatePassword(password: string): PasswordValidationStatus;
```

Comment:
There will be 2 private helper methods (_checkLengthOptions` and `_checkCharacterOptions`) defined in `PasswordPolicyImpl` which will implement this interface. I did not include this class in this PR to make this easier to review.

Reply:
should we add public static annotation?

Reply:
It won't be static since this method needs access to the internal fields (requirements) of the policy.